// fusion-plugin-styletron-react/dist-browser-cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "styled", {
  enumerable: true,
  get: function () {
    return _styletronReact.styled;
  }
});
Object.defineProperty(exports, "withStyle", {
  enumerable: true,
  get: function () {
    return _styletronReact.withStyle;
  }
});
Object.defineProperty(exports, "withStyleDeep", {
  enumerable: true,
  get: function () {
    return _styletronReact.withStyleDeep;
  }
});
Object.defineProperty(exports, "withTransform", {
  enumerable: true,
  get: function () {
    return _styletronReact.withTransform;
  }
});
Object.defineProperty(exports, "withWrapper", {
  enumerable: true,
  get: function () {
    return _styletronReact.withWrapper;
  }
});
Object.defineProperty(exports, "AtomicPrefixToken", {
  enumerable: true,
  get: function () {
    return _constants.AtomicPrefixToken;
  }
});
exports.default = void 0;

var _styletronReact = require("styletron-react");

var _constants = require("./constants.js");

var _browser = _interopRequireDefault(require("./browser"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var _default = false ? server : _browser.default;

exports.default = _default;
